File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
import sys

def solution(gems):
    
    answer = [1, len(gems)]
    
    INF = sys.maxsize
    MIN = INF
    n = len(gems)
    gems_set = set(gems)
    end = 0 
    
    gem_dict = {}
    
    gem_dict[gems[0]] = 1 
    
    for start in range(n) :
        while end < n and len(gem_dict) < len(gems_set) :
            end += 1
            
            if end != n :
                
                if gems[end] not in gem_dict :
                    gem_dict[gems[end]] = 1
                else :
                    gem_dict[gems[end]] += 1
        
        if end == n :
            break
        
        if end - start < answer[1] - answer[0] :
            answer = [start + 1, end + 1]
            
        gem_dict[gems[start]] -= 1
        if gem_dict[gems[start]] == 0 :
            del gem_dict[gems[start]]
      
    return answer

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;

class Solution {
    public int[] solution(String[] gems) {
        int[] answer = new int[2];
        
        int gemsCnt = new HashSet<>(Arrays.asList(gems)).size();
        int length = 1000001;
        
        Map<String, Integer> gemsMap = new HashMap<>();
        int end = 0;
        
        gemsMap.put(gems[0], 1);
        for (int start = 0; start < gems.length; start++) {
            while (end < gems.length && gemsMap.size() < gemsCnt) {
                end += 1;
                if (end != gems.length) gemsMap.put(gems[end], gemsMap.getOrDefault(gems[end], 0 ) + 1);
            }
            
            if (end == gems.length) break;
            
            if (end - start + 1 < length) {
                length = end - start + 1;

                answer[0] = start + 1;
                answer[1] = end + 1;
            }
          
            
            gemsMap.put(gems[start], gemsMap.get(gems[start]) - 1);
            
            if (gemsMap.get(gems[start]) == 0)
                gemsMap.remove(gems[start]);
        }
        return answer;
    }
}